File: Leetcode/86-Partition-List/PartitionList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} x
 * @return {ListNode}
 */
var partition = function(head, x) {
    var cur = head
    var leftDummy = new ListNode(0)
    var leftCur = leftDummy
    var rightDummy = new ListNode(0)
    var rightCur = rightDummy
    while (cur) {
        if (cur.val<x) {
            leftCur.next = cur
            cur=cur.next
            leftCur=leftCur.next
            leftCur.next = null
        } else {
            rightCur.next = cur
            cur = cur.next
            rightCur = rightCur.next
            rightCur.next = null
        }
    }
    leftCur.next = rightDummy.next
    return leftDummy.next
};

File: Leetcode/86-Partition-List/PartitionList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class PartitionList {
    public ListNode partition(ListNode head, int x) {
        // ListNode dummy = new ListNode(0);
        // dummy.next = head;
        ListNode cur = head;
        
        ListNode leftDummy = new ListNode(0);
        ListNode leftCur = leftDummy;
        
        ListNode rightDummy = new ListNode(0);
        ListNode rightCur = rightDummy;
        
        while (cur!=null) {
            if (cur.val < x) {
                leftCur.next = cur;
                cur = cur.next;
                leftCur = leftCur.next;
                leftCur.next = null;
            } else {
                rightCur.next = cur;
                cur = cur.next;
                rightCur = rightCur.next;
                rightCur.next = null;
            }
        }
        leftCur.next = rightDummy.next;
        return leftDummy.next;
    }
}




